File: tests/refs/binary_search_on_answer.py
"""Reference: binary-search-on-answer (Koko eating bananas, LC 875)."""

from __future__ import annotations
from typing import List


def _can_finish(piles: List[int], h: int, speed: int) -> bool:
    hours = 0
    for pile in piles:
        hours += (pile + speed - 1) // speed
    return hours <= h


def minEatingSpeed(piles: List[int], h: int) -> int:
    lo, hi = 1, max(piles)
    while lo < hi:
        mid = lo + (hi - lo) // 2
        if _can_finish(piles, h, mid):
            hi = mid
        else:
            lo = mid + 1
    return lo

File: tests/refs/BinarySearchOnAnswer.java
/*
 * Reference Java implementation for tests/cases/binary_search_on_answer.json.
 * Koko eating bananas: binary search on the minimum speed.
 */
public class BinarySearchOnAnswer {
    private static boolean canFinish(long[] piles, long h, long speed) {
        long hours = 0;
        for (long pile : piles) {
            hours += (pile + speed - 1) / speed;
        }
        return hours <= h;
    }

    public static long minEatingSpeed(long[] piles, long h) {
        long lo = 1, hi = 0;
        for (long p : piles) if (p > hi) hi = p;
        while (lo < hi) {
            long mid = lo + (hi - lo) / 2;
            if (canFinish(piles, h, mid)) hi = mid;
            else lo = mid + 1;
        }
        return lo;
    }
}
